ykrt/src/compile/mod.rs

```diff
@@ -38,15 +52,35 @@ struct SendSyncConstPtr<T>(*const T);
 unsafe impl<T> Send for SendSyncConstPtr<T> {}
 unsafe impl<T> Sync for SendSyncConstPtr<T> {}
 
-struct Guard {
-    failed: u32,
-    code: Option<SendSyncConstPtr<c_void>>,
+pub struct Guard {
+    failed: AtomicU32,
+    ct: Mutex<Option<Arc<CompiledTrace>>>,
+}
+
+impl Guard {
+    pub fn inc(&self) {
+        let n = self.failed.load(Ordering::Relaxed);
```

Comment:
This has a race condition: you could load "3", another thread can store "4", then you store "4" i.e. you've lost an increment. You want `fetch_add`, which avoids this race.

Reply:
Will fix. But this made me realise if two threads have the same control point, all the hot counts are gonna be shared, so we'll reach the hot threshold much faster. Not saying this is a problem, just something I hadn't thought about before.

Reply:
Fixed in b009e68cd

Reply:
> But this made me realise if two threads have the same control point, all the hot counts are gonna be shared, so we'll reach the hot threshold much faster.

Yep. It's not a problem though!